projects/transformers/callbacks/metrics.py

```diff
@@ -72,11 +78,19 @@ def on_evaluate(self, args, state, control, metrics, **kwargs):
             self.eval_metrics["num_nonzero_params"].append(num_nonzero)
             self.eval_metrics["sparsity"].append(model_sparsity)
 
+            # guarantee that everything stayed sparse,
+            # up to specified tolerance
+            if (self.sparsity_tolerance < 1) and len(self.eval_metrics["sparsity"]) > 1:
```

Comment:
This is mainly to catch the bug where RezeroWeights callback is not in use, and the model starts un-zeroing weights. There is another explicit check for the callback elsewhere, but I wanted to have another way of avoiding unexpected changes in sparsity. We could change it to a warning if you prefer.

Reply:
An assert would cause the whole experiment to fail, and towards the end of training. Ideally, if something fails, it should be at the start of a run.

Reply:
That depends on what eval_steps is set to. I made the assert assuming if this callback is in place, you want to evaluate periodically, in which case it will fail at the first eval. Given that to get here, you needed to pass check_sparsity_callback ( afunction that also looks to ensure RezeroWeights is on), I think this is  not likely to fail at the end too often. But that is a guess after all. I'll go ahead and switch it to warning (in the next PR).